Lib/test/test_ctypes/test_win32_com_foreign_func.py

```diff
@@ -0,0 +1,205 @@
+import ctypes
+import gc
+import sys
+import unittest
+from ctypes import POINTER, byref, c_void_p
+from ctypes.wintypes import BYTE, DWORD, WORD
+
+COINIT_APARTMENTTHREADED = 0x2
+CLSCTX_SERVER = 5
+S_OK = 0
+OUT = 2
+TRUE = 1
+E_NOINTERFACE = -2147467262
+
+PyInstanceMethod_New = ctypes.pythonapi.PyInstanceMethod_New
+PyInstanceMethod_New.argtypes = [ctypes.py_object]
+PyInstanceMethod_New.restype = ctypes.py_object
+PyInstanceMethod_Type = type(PyInstanceMethod_New(id))
+
+
+class GUID(ctypes.Structure):
+    # https://learn.microsoft.com/en-us/windows/win32/api/guiddef/ns-guiddef-guid
+    _fields_ = [
+        ("Data1", DWORD),
+        ("Data2", WORD),
+        ("Data3", WORD),
+        ("Data4", BYTE * 8),
+    ]
+
+
+class ProtoComMethod:
+    """Utilities for tests that define COM methods."""
+
+    def __init__(self, index, restype, *argtypes):
+        self.index = index
+        self.proto = ctypes.WINFUNCTYPE(restype, *argtypes)
+
+    def __set_name__(self, owner, name):
+        foreign_func = self.proto(self.index, name, *self.args)
+        self.mth = PyInstanceMethod_Type(foreign_func)
+
+    # NOTE: To aid understanding, adding type annotations with `typing`
+    # would look like this.
+    #
+    # _ParamFlags = tuple[tuple[int] | tuple[int, str] | tuple[int, str, Any], ...]
+    #
+    # @overload
+    # def __call__(self, paramflags: None | _ParamFlags, iid: GUID, /) -> Self: ...
+    # @overload
+    # def __call__(self, paramflags: None | _ParamFlags, /) -> Self: ...
+    # @overload
+    # def __call__(self) -> Self: ...
+    def __call__(self, *args):
+        self.args = args
+        return self
+
+    def __get__(self, instance, owner=None):
+        # if instance is None:
+        #     return self
+        # NOTE: In this test, there is no need to define behavior for the
+        # custom descriptor as a class attribute, so the above implementation
+        # is omitted.
+        return self.mth.__get__(instance)
```

Comment:
This is a very convoluted way of collecting the necessary arguments. IMO, it distracts from the point of the test. As far as I can see, `ProtoComMethod` can be replaced with:
```py
def ProtoComMethod(name, index, restype, *argtypes):
    proto = ctypes.WINFUNCTYPE(restype, *argtypes)
    def make_method(*args):
        foreign_func = proto(index, name, *args)
        def call(self, *args, **kwargs):
            return foreign_func(self, *args, **kwargs)
        return call
    return make_method
```
with the difference in usage being that `name` needs to be passed in explicitly, as in:
```diff
     proto_query_interface = ProtoComMethod(
+       'QueryInterface',
         0, HRESULT, POINTER(GUID), POINTER(c_void_p)
     )
```
This way the `instancemethod` type (`PyInstanceMethod_New` and `PyInstanceMethod_Type` above) isn't needed.

Reply:
I thought it was a cool idea to use the descriptor protocol to avoid having to specify a string with the same name as the attribute.  
But you're right—this is test code, not the package's production code. As you pointed out, even if it's redundant, keeping it as simple as possible is better.

Reply:
> This way the `instancemethod` type (`PyInstanceMethod_New` and `PyInstanceMethod_Type` above) isn't needed.

To be honest, I didn’t really understand why `instancemethod` was necessary in the previous implementation.
I just reused the implementation used in `comtypes`.